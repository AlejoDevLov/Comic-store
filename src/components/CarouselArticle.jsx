import { Link } from "react-router-dom"

export const CarouselArticle = ({ propsCarousel }) => {

    const {divGrande, elementsCarousel, setClassIconArrow,
            classIcon,setClassIconArrow2, classIcon2,
            removeClassIconArrow, removeClassIconArrow2, translateCarousel} 
        = propsCarousel;

  return (
    <section className="related-products">

          <h2>Productos similares</h2>

          <div className="slider-container">
            <div className="slider-grande" ref={divGrande}>
              {
                elementsCarousel.map( (item, i) => (
                  <div className="item-carousel" key={i}>
                    <Link to={`/articlePage/${item.id}-${item.name.replace(/\s/g,'-')}`}><img src={item.url} alt={item.name} /></Link>
                    <h3>{item.name}</h3>
                  </div>
                ))
              }
            </div>
          </div>

          <div className="arrows-slider">
            <div 
              onMouseDown={setClassIconArrow} 
              onMouseUp={() => (removeClassIconArrow(), translateCarousel())}
              className={classIcon}
            >
              <ion-icon name="arrow-back-circle"></ion-icon>
            </div>
            <div 
              onMouseDown={setClassIconArrow2} 
              onMouseUp={ () => (removeClassIconArrow2(), translateCarousel())}
              className={classIcon2}
            >
              <ion-icon name="arrow-forward-circle"></ion-icon>
            </div>
          </div>

        </section>
  )
}
