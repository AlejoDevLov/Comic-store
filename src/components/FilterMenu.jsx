import { useContext, useEffect, useRef } from "react"
import { allAnimes } from "../helpers";
import { listContext } from "../states/shopListState/listContext";

export const FilterMenu = ({ refClass, classFilter }) => {

    const { setProducts, initialProducts } = useContext(listContext);

    const arrayAnime = [...allAnimes];


    const anime = useRef();
    const product = useRef();
    const price = useRef();


    const setFilters = () => {

        if( anime.current.value == undefined || product.current.value == undefined || price.current.value == undefined) return;
    
        if( anime.current.value !== '' && product.current.value !== '' && price.current.value !== '' ){
                setProducts( arrayAnime.filter( item => (
                item.anime === anime.current.value
            )).filter( item => item.keyWords.includes(product.current.value)).sort( (a,b) => {
                if( price.current.value === 'menor' ){
                    return Number(a.precio.replace(/[^0-9.-]+/g,"")) - Number(b.precio.replace(/[^0-9.-]+/g,""))
                } else{
                    return Number(b.precio.replace(/[^0-9.-]+/g,"")) - Number(a.precio.replace(/[^0-9.-]+/g,""))
                }
            }));
            return;
        }

        if( anime.current.value !== '' && product.current.value !== '' ){
            setProducts( arrayAnime.filter( item => item.anime === anime.current.value).filter( item => item.keyWords.includes(product.current.value)));
            return;
        }

        if( anime.current.value !== '' && price.current.value !== '' ){
            setProducts( arrayAnime.filter( item => item.anime === anime.current.value).sort( (a,b) => {
                if( price.current.value === 'menor' ){
                    return Number(a.precio.replace(/[^0-9.-]+/g,"")) - Number(b.precio.replace(/[^0-9.-]+/g,""))
                } else{
                    return Number(b.precio.replace(/[^0-9.-]+/g,"")) - Number(a.precio.replace(/[^0-9.-]+/g,""))
                }
            }));
            return;
        }

        if( product.current.value !== '' && price.current.value !== ''){
            setProducts( [...initialProducts].filter( item => item.keyWords.includes(product.current.value)).sort( (a,b) => {
                if( price.current.value === 'menor' ){
                    return Number(a.precio.replace(/[^0-9.-]+/g,"")) - Number(b.precio.replace(/[^0-9.-]+/g,""))
                } else{
                    return Number(b.precio.replace(/[^0-9.-]+/g,"")) - Number(a.precio.replace(/[^0-9.-]+/g,""))
                }
            }));
            return;
        }

        if( anime.current.value !== '' ){
            setProducts( arrayAnime.filter( item => item.anime === anime.current.value));
            setInitialProducts( arrayAnime.filter( item => item.anime === anime.current.value));
        } 
        if( product.current.value !== '') setProducts( initialProducts.filter( item => item.keyWords.includes(product.current.value)));
        if( price.current.value !== '' ){
            setProducts([...initialProducts].sort( (a,b) => {
                if( price.current.value === 'menor' ){
                    return Number(a.precio.replace(/[^0-9.-]+/g,"")) - Number(b.precio.replace(/[^0-9.-]+/g,""))
                } else{
                    return Number(b.precio.replace(/[^0-9.-]+/g,"")) - Number(a.precio.replace(/[^0-9.-]+/g,""))
                }
            }))
        } 
    }

    const cleanFilters = () => {
        anime.current.value = '';
        product.current.value = '';
        price.current.value = '';

        setProducts(initialProducts);
        
    }


  return (
    <>
        <div className={classFilter} ref={ refClass }>

            <div className="filter-anime">
                <label htmlFor="anime-select">Anime/Caricatura</label>
                <select name="anime" id="anime-select" ref={anime}>
                    <option value=""></option>
                    <option value="Dragon ball">Dragon Ball</option>
                    <option value="demon slayer">Demon Slayer</option>
                    <option value="seven deadly">Nanatsu no Taizai</option>
                    <option value="los Simpson">Los Simpsons</option>
                    <option value="Mario">Mario Bros</option>
                </select>
            </div>

            <div className="filter-product">
                <label htmlFor="product-select">Producto</label>
                <select name="product" id="product-select" ref={product}>
                    <option value=""></option>
                    <option value="camiseta">Camisetas</option>
                    <option value="llavero">Llaveros</option>
                    <option value="vaso">Vasos</option>
                    <option value="case">Cases</option>
                    <option value="figura">Figuras</option>
                </select>
            </div>

            <div className="filter-price">
                <label htmlFor="price-select">Precios</label>
                <select name="price" id="price-select" ref={price}>
                    <option value=""></option>
                    <option value="menor">Menor a mayor</option>
                    <option value="mayor">Mayor a menor</option>
                </select>
            </div>
            
            <div className="buttons-filter">
                <button onClick={setFilters}>Aplicar filtros</button>
                <button onClick={cleanFilters}>Limpiar filtros</button>
            </div>
        </div>
    </>
  )
}
