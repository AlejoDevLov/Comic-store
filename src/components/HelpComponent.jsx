
export const HelpComponent = ({ helpIcon, classIcon }) => {

    
  return (
    <>
        <div className={classIcon} ref={helpIcon}>
            <h4>¿Necesitas ayuda?</h4>
            <h4>Sobre Nosotros</h4>
            <h4>Contáctanos</h4>
        </div>
    </>
  )
}
