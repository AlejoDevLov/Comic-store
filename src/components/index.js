

export * from './Header';
export * from './Shop';
export * from './MenuCarList';
export * from './Article';
export * from './FilterMenu';
export * from './SizeTShirt';
export * from './HelpComponent';
export * from './IconsNavbar';