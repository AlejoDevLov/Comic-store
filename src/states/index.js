

export * from './carState';
export * from './listCarState';
export * from './shopListState';