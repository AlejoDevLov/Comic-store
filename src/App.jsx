import { Navigate, Route, Routes } from 'react-router-dom';
import HomePage from './pages/HomePage';
import { ShopPage, ArticlePage  } from './pages';
import { CarProvider } from './states/carState';


export const App = () => {
  return (
    <CarProvider>
      <Routes>
        <Route path='/home' element={ <HomePage /> }/>      
        <Route path='/*' element={ <Navigate to="/home"/> }/>
        <Route path='/shop/:anime' element={ <ShopPage/> } />
        <Route path='/articlePage/:itemName' element={ <ArticlePage /> }/>
      </Routes>
    </CarProvider>
  )
}
