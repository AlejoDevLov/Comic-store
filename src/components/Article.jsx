import { useContext, useEffect, useState } from "react";
import { carContext } from "../states/carState";
import { SizeTShirt } from "./SizeTShirt";

const initialClass = '';

export const Article = ({ element, onSetState }) => {

  const [classBtn, setClassBtn] = useState(initialClass);

  const { url, id, name, precio, anime, cantidad } = element;

  const { dispatchCar } = useContext(carContext)
  
  const addArticle = () => {
    const action = {
      type: '[TODO] add article',
      payload: element,
    }

    dispatchCar(action);
  }

  const executeFn = (anime, id) => {
    addArticle();
    onSetState(anime, id)
  }

  const isAvailable = ( cantidad < 1 );

  useEffect( () => {
    isAvailable ? setClassBtn('disabled') : setClassBtn('');
  }, [isAvailable] );


  return (
      <div className="card animate__animated animate__slideInUp" key={ id }>
          <div className='image'><img src={url} alt={anime} /></div>
          <div className='title'><h2>{name}</h2></div>
            {
              element?.talla && <SizeTShirt sizes={ element.talla }/> 
            }
          <div className='price'>{precio}</div>
          <div className="cantidad">Cantidad disponible: <span className="color-cantidad">{ cantidad }</span></div>
          <div className='button'>
              <button 
                disabled={ isAvailable } 
                className={ classBtn }
                onClick={ () => executeFn(anime,id) }
                >Comprar
              </button>
          </div>
      </div>
  )
}
