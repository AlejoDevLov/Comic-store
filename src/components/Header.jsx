import { Link } from 'react-router-dom';
import '../CSS/header.css';
import { carContext } from '../states/carState';
import { useContext, useEffect, useRef, useState } from 'react';
import { carListContext } from '../states/listCarState';
import { listContext } from '../states/shopListState/listContext';
import { FilterMenu } from './FilterMenu';
import { HelpComponent } from './HelpComponent';


export const Header = () => {

  const { stateCar } = useContext( carContext );
  const { handleCarList } = useContext(carListContext);
  const { setProducts, initialProducts, products } = useContext(listContext);

  const [classCarDiv, setClassCarDiv] = useState('car-counter');
  const [classFilter, setClassFilter] = useState('filter-container');
  const [classHelpIcon, setClassHelpIcon] = useState('container-help');

  const inputSearch = useRef();
  const iconFilter = useRef();
  const helpIcon = useRef();

  useEffect( () => {
    stateCar.length == 0 && setClassCarDiv('car-counter hide-counter');
  },[]);

  useEffect( () => {
    stateCar.length > 0 
      ? setClassCarDiv('car-counter') 
      : setClassCarDiv('car-counter hide-counter');
  },[stateCar]);

  const totalElemetnsCar = stateCar.reduce( (prevValue, item) => {
      return prevValue + item.units;
  }, 0); 

  const searchKeywords = (e) => {
    // console.log(e.key);
    if(e.key == 'Enter'){
      const keys = e.target.value.toLowerCase().trim();
      if( keys != ''){
        setProducts( initialProducts.filter( product => {
          return product.keyWords.includes(keys);
        }) )
        return;
      }
      setProducts(initialProducts);
    }
  }

  const searchKeywordsOnClick = () => {
    const keys = inputSearch.current.value.toLowerCase().trim();
    if( keys != ''){
      setProducts( initialProducts.filter( product => {
          return product.keyWords.includes(keys);
      }) )
      return;
    }
    setProducts(initialProducts);
  }

  const setFilterClass = () => {
    if(!iconFilter) return;
    if( iconFilter.current.classList.contains('filter-clip')){
      setClassFilter('filter-container');
    } else{
      setClassFilter('filter-container filter-clip');
    }
  }

  const setIconHelp = () => {
    if(helpIcon.current.classList.contains('filter-clip')){
      setClassHelpIcon('container-help');
    }else{
      setClassHelpIcon('container-help filter-clip');
    }
  }


  return (
    <>
        <nav className="navbar">
            <div className='filter'><img src="../../icons/Bars3BottomLeft.svg" alt="menu-icon" height="50%" onClick={setFilterClass}/></div>
            <div className="search"><input type="text" placeholder='Buscar...' onKeyDown={ searchKeywords } ref={inputSearch}/><img src="../../icons/MagnifyingGlass.svg" alt="icon-search" height="50%" onClick={ searchKeywordsOnClick }/></div>
            <div className="icons">
                <div className="home"><Link to='/home' title='Home'><img src="../../icons/Home.svg" alt="home icon" height="50%"/></Link></div>
                <div className="car">
                    <img src="../../icons/ShoppingCart.svg" alt="icon car" height="50%" onClick={handleCarList}/>
                  <div className={classCarDiv} >{totalElemetnsCar}</div>
                </div>
                <div className="account"><img src="../../icons/help-circle-outline.svg" alt="user-icon" height="50%" onClick={setIconHelp}/></div>
            </div>
        </nav>
        <FilterMenu refClass={iconFilter} classFilter={classFilter}/>
        <HelpComponent helpIcon={helpIcon} classIcon={classHelpIcon}/>
    </>
  )
}
