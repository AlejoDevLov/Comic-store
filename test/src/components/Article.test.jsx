import { fireEvent, render, screen } from "@testing-library/react"
import { Article } from "../../../src/components/Article"
import { carContext } from "../../../src/states";
import { MemoryRouter } from "react-router-dom";


describe('Pruebas en Article', () => {
    
    const item = {
        id: 4,
        name: 'Camiseta Vegeta',
        descripcion: 'Camiseta  hecha en Algodon 80% y PoliÃ©ster 20%. Con acabados de alta calidad. Producto 100% Colombiano.',
        keyWords: ['camiseta', 'dragon ball', 'vegeta', 'modo dios'],
        talla: ['S', 'M'],
        precio: '$50.000',
        colores: [ 'Negra', 'Gris'],
        anime: 'Dragon ball',
        cantidad: 3,
        url: '/images/dragon-ball/camisetas/4.png',
    }

    const onSetState = jest.fn();
    const state = [];
    const dispatchCar = jest.fn();

    test('Debe hacer match con el snapshop', () => {
        
        const {container} = render(
            <carContext.Provider value={{ state, dispatchCar }}>
                <MemoryRouter>
                    <Article element={item} onSetState={onSetState}/>
                </MemoryRouter>
            </carContext.Provider>
        )
        // screen.debug();

        expect(container).toMatchSnapshot('Article');
    });

    test('Debe mostrar la informacion del item', () => {
        render(
            <carContext.Provider value={{ state, dispatchCar }}>
                <MemoryRouter>
                    <Article element={item} onSetState={onSetState}/>
                </MemoryRouter>
            </carContext.Provider>
        )

        expect(screen.getByText('Camiseta Vegeta')).toBeTruthy();
    });

    test('Debe llamar la funcion executeFn al dar click en btn', () => {        

        render(
            <carContext.Provider value={{ state, dispatchCar }}>
                <MemoryRouter>
                    <Article element={item} onSetState={onSetState}/>
                </MemoryRouter>
            </carContext.Provider>
        );
        const btn = screen.getByRole('button');
        fireEvent.click(btn);
        // se ejecuta console.log desde la fn en useArticle.js
    })
})