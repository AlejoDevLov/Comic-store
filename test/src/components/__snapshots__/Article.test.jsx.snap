// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Article Debe hacer match con el snapshop: Article 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="image"
    >
      <a
        href="/articlePage/4-Camiseta-Vegeta"
      >
        <img
          alt="Dragon ball"
          loading="lazy"
          src="/images/dragon-ball/camisetas/4.png"
        />
      </a>
    </div>
    <div
      class="title"
    >
      <h2>
        Camiseta Vegeta
      </h2>
    </div>
    <div
      class="size-container"
    >
      <label
        for="talla"
      >
        Talla: 
      </label>
      <select
        data-testid="select"
        id="talla"
        name="talla"
      >
        <option
          value=""
        />
        <option
          value="S"
        >
          S
        </option>
        <option
          value="M"
        >
          M
        </option>
      </select>
    </div>
    <div
      class="price"
    >
      $50.000
    </div>
    <div
      class="cantidad"
    >
      Cantidad disponible: 
      <span
        class="color-cantidad"
      >
        3
      </span>
    </div>
    <div
      class="button"
    >
      <button
        class=""
      >
        Comprar
      </button>
    </div>
  </div>
</div>
`;
